bech32/app/Main.hs

```diff
@@ -0,0 +1,142 @@
+    ( bitcoinAlphabet, decodeBase58, unAlphabet )
+import Data.Char
+    ( isLetter, isLower, isUpper, ord, toLower )
+import Data.Maybe
+    ( fromJust )
+import Options.Applicative
+    ( Parser
+    , ParserInfo
+    , argument
+    , customExecParser
+    , eitherReader
+    , help
+    , helper
+    , info
+    , metavar
+    , optional
+    , prefs
+    , progDesc
+    , showHelpOnEmpty
+    , (<|>)
+    )
+import System.IO
+    ( BufferMode (..), Handle, hSetBuffering, stderr, stdin, stdout )
+
+import qualified Codec.Binary.Bech32 as Bech32
+import qualified Data.ByteArray.Encoding as BA
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+
+
+main :: IO ()
+main = setup >> parse >>= run
+
+newtype Cmd = Cmd
+  { prefix :: Maybe HumanReadablePart
+  } deriving (Show)
+
+-- | Enable ANSI colors on Windows and correct output buffering
+setup :: IO ()
+setup =
+    mapM_ hSetup [stderr, stdout]
+  where
+    hSetup :: Handle -> IO ()
+    hSetup h = hSetBuffering h NoBuffering
+
+-- | Parse command line options and arguments
+parse :: IO Cmd
+parse = customExecParser (prefs showHelpOnEmpty) parser
+  where
+    parser :: ParserInfo Cmd
+    parser = info (helper <*> cmd) $ mconcat
+        [ progDesc "Simple utility for converting to/from bech32 byte strings."
+        ]
+
+    cmd :: Parser Cmd
+    cmd = Cmd <$> optional hrpArgument
+
+-- | Parse a 'HumanReadablePart' as an argument.
+hrpArgument :: Parser HumanReadablePart
+hrpArgument = argument (eitherReader reader) $ mempty
+    <> metavar "PREFIX"
+    <> help "A human-readable prefix (e.g. 'addr')."
+  where
+    reader :: String -> Either String HumanReadablePart
+    reader = left show . humanReadablePartFromText . T.pack
+
+-- | Run a Command in IO
+run :: Cmd -> IO ()
+run Cmd{prefix} = do
+    source <- T.decodeUtf8 . B8.filter (/= '\n') <$> B8.hGetContents stdin
+    case prefix of
+        Nothing  -> runDecode source
+        Just hrp -> runEncode hrp source
+  where
+    runDecode source =
+        case Bech32.decodeLenient source of
+          Left err ->
+              fail (show err)
+          Right (_, dataPart) ->
+              B8.putStrLn $ convertToBase BA.Base16 $ fromJust $
+                dataPartToBytes dataPart
+
+    runEncode hrp source = do
+        datapart <- either fail pure $
+            case detectEncoding (T.unpack source) of
+                Just Base16 -> do
+                    let fromBase16 = convertFromBase BA.Base16 . T.encodeUtf8
+                    dataPartFromBytes <$> fromBase16 source
+                Just Bech32 ->
+                    right snd $ left show $ Bech32.decodeLenient source
+                Just Base58 -> do
+                    let maybeToEither = maybe (Left "Invalid Base58-encoded string.") Right
+                    let fromBase58 = decodeBase58 bitcoinAlphabet . T.encodeUtf8
+                    dataPartFromBytes <$> maybeToEither (fromBase58 source)
```

Comment:
The function is so simple here that it feels wrong importing a dependency / library just for this.

Reply:
> The function is so simple here that it feels wrong importing a dependency / library just for this.

We do already import this dependency (for another target), so it would not harm our build times to import it here.

In any case, I would argue that the [`maybeToEither`](https://hackage.haskell.org/package/extra-1.7.3/docs/Data-Either-Extra.html#v:maybeToEither) function is [widely used](https://packdeps.haskellers.com/reverse/extra). Even when people redefine it, they typically give it the same type! [`:: a -> Maybe b -> Either a b`](https://hoogle.haskell.org/?hoogle=maybeToEither).

For that reason, I think it's not a bad idea to follow this convention.